### PR TITLE
[REVIEW] Fix `DatetimeIndex` & `TimedeltaIndex` constructors

### LINKED ISSUE #11335: DatetimeIndex yielding AttributeError
**Describe the bug**
Having been trying to extract the month, year, dayofweek etc from a cudf.DatetimeIndex object, I get the `'StringColumn' object has no attribute 'get_dt_field'` error.

**Steps/Code to reproduce bug**
```
import cudf
dic = {'time' : ["2015-01-15 19:05:39"]}
cudf_df = cudf.DataFrame(dic)
x =  cudf_df["time"]
x_date = cudf.DatetimeIndex(x)
x_date.month
```
**Expected behavior**
Output should be `Int16Index([1], dtype='int16')`

**Environment overview**
-Environment location: Jupyter Notebook
-Method of cuDF install : conda
@AjayThorve

### PR DESCRIPTION
**Pull request by galipremsagar**
## Description
Closes #11335 

This PR fixes an issue with `DatetimeIndex` & `TimedeltaIndex` where the underlying columns would still be of numeric or string types rather than `DatetimeColumn` and `TimedeltaColumn` respectively. This is the actual root cause that leads to errors in some downstream API calls.

## Checklist
- [x] I am familiar with the [Contributing Guidelines](https://github.com/rapidsai/cudf/blob/HEAD/CONTRIBUTING.md).
- [x] New or existing tests cover these changes.
- [x] The documentation is up to date with these changes.

### CODE REVIEW

File: python/cudf/cudf/core/index.py [hunk 1]
```diff
@@ -1784,14 +1784,18 @@ def __init__(
...
         if isinstance(data, np.ndarray) and data.dtype.kind == "M":
             data = column.as_column(data)
         elif isinstance(data, pd.DatetimeIndex):
             data = column.as_column(data.values)
         elif isinstance(data, (list, tuple)):
             data = column.as_column(np.array(data, dtype=dtype))
+        elif isinstance(data, cudf.Series):
+            data = data._column.astype(dtype)
+        elif isinstance(data, column.ColumnBase):
+            data = data.astype(dtype)
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/index.py:1798]: I feel like some of this logic that handles separate objects is already in `as_column`. How much of this still works if we just replace it with `data = as_column(data, dtype=dtype)`? Ideally we'd just have one funnel for this.

File: python/cudf/cudf/core/index.py [hunk 2]
```diff
@@ -2263,15 +2267,25 @@ def __init__(
                 "dtype parameter is supported"
             )
 
+        valid_dtypes = tuple(
+            f"timedelta64[{res}]" for res in ("s", "ms", "us", "ns")
+        )
+        if dtype not in valid_dtypes:
+            raise TypeError("Invalid dtype")
```
Comment (brandon-b-miller) [inline python/cudf/cudf/core/index.py:2274]: same comment as above for having this be in `as_column`

File: python/cudf/cudf/core/index.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: Looks like `as_column` will handle everything that was already here. So I just deleted all the duplication here as re-directed everything to `as_column`.

File: python/cudf/cudf/core/index.py (hunk 2 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py:2267]: Addressed similar to above.

File: python/cudf/cudf/core/index.py (hunk 1 above)
Comment (galipremsagar) [reply python/cudf/cudf/core/index.py]: cc:@vyasr just FYI

Comment (vyasr) [reply python/cudf/cudf/core/index.py]: Yup I think that's fine. In the long run I anticipate that `as_column` (or a cleaned up version) will be used for constructors, but all other APIs will be much more prescriptive in what they accept. We can worry about `as_column` then, for now it's fine to increase our dependence on it.

**Review by vyasr — APPROVED** [review]

**Comment by galipremsagar** [comment]
@gpucibot merge

**Pull request merged**